src/webview/ExtensionToWebviewMessageSender.ts

```diff
@@ -0,0 +1,105 @@
+import * as vscode from "vscode";
+import { ChatWebviewProvider } from "./ChatWebviewProvider";
+import {
+  ExtensionToWebviewMessageType,
+  ExtensionToWebviewMessageTypeParamsMap,
+  ExtensionToWebviewMessage,
+  Chat,
+} from "../../shared/types";
+import { ChatDataManager } from "../Chat/ChatDataManager";
+
+export class ExtensionToWebviewMessageSender {
+  private readonly chatWebviewProvider: ChatWebviewProvider;
+  private readonly chatDataManager: ChatDataManager;
+
+  public constructor(
+    chatWebviewProvider: ChatWebviewProvider,
+    chatDataManager: ChatDataManager,
+  ) {
+    this.chatWebviewProvider = chatWebviewProvider;
+    this.chatDataManager = chatDataManager;
+  }
+
+  /**
+   * Refreshes the webview to reflect the current state of the chat manager, including
+   * the list of chats and the active chat.
+   * Requests the webview to be shown.
+   */
+  public async refresh() {
+    await this.chatWebviewProvider.show();
+    const chats: Chat[] = this.chatDataManager.chats;
+    let activeChatId: number | null = this.chatDataManager.activeChatId;
+    if (activeChatId && !chats.find((chat) => chat.id === activeChatId)) {
+      // activeChatId is invalid, so clear it
+      this.chatDataManager.activeChatId = null;
+      activeChatId = null;
+      return;
+    }
+
+    const messageType: ExtensionToWebviewMessageType = "refresh";
+    await this.postMessage({
+      messageType: messageType,
+      params: {
+        chats: chats,
+        activeChatId: activeChatId,
+      } as ExtensionToWebviewMessageTypeParamsMap[typeof messageType],
+    } as ExtensionToWebviewMessage);
+  }
+
+  /**
+   * Updates the webview to display an error message. Shows error notification if the webview is not visible.
+   * @param errorMessage The error message to display.
+   */
+  public async updateErrorMessage(errorMessage: string) {
+    if (!this.chatWebviewProvider.isWebviewVisible) {
+      vscode.window.showErrorMessage(errorMessage);
+    } else {
+      const messageType: ExtensionToWebviewMessageType = "errorMessage";
+      await this.postMessage({
+        messageType: messageType,
+        params: {
+          errorMessage: errorMessage,
+        } as ExtensionToWebviewMessageTypeParamsMap[typeof messageType],
+      } as ExtensionToWebviewMessage);
+    }
+  }
+
+  /**
+   * Updates the webview to reflect whether or not a message is currently loading.
+   * @param isLoading If the message is loading.
+   */
+  public async updateIsMessageLoading(isLoading: boolean) {
```

Comment:
I like the idea of displaying "typing" or something similar in the UI.

Reply:
It controls the display of a loading indicator at the bottom of the chat when the user sends a message (or the "review code"/"explain code" commands have caused a message to be sent) and is waiting on a response.

In a later PR, I was thinking of tying a "loading"/"busy" state to a specific chat ID. As it is implemented, there is only one loading state on the front end. So currently, a user could send  message in one conversation (which will then show as loading) and then navigate to a different chat, but that chat will appear as if it is also has a message loading and waiting for a response.

Would you prefer text like "Typing..." or "Thinking..." instead of the spinner? Or perhaps in addition to where there is a small spinner next to the text.

Reply:
I like the idea of using text instead of the spinner. It would be nice if there were still some "movement" like having the dots appear and disappear in sequence.